api/applications_api.py

```diff
@@ -57,39 +43,97 @@ def perform_create(self, serializer):
 
 
 class OrganisationSerializer(serializers.ModelSerializer):
+
+    id = serializers.IntegerField(allow_null=True, required=False)
+
+    def to_internal_value(self, data):
+        if "id" not in data or data["id"] is None:
+            return super(self.__class__, self).to_internal_value(data)
+        if "id" in self.fields:
+            try:
+                obj_id = self.fields["id"].to_internal_value(data["id"])
+            except ValidationError as exc:
+                raise ValidationError({"id": exc.detail})
+            for field in self.fields.values():
+                for validator in field.validators:
+                    if type(validator) == UniqueValidator and obj_id is not None:
+                        # Exclude id from queryset for checking uniqueness
+                        validator.queryset = validator.queryset.exclude(id=obj_id)
+        return super(self.__class__, self).to_internal_value(data)
+
     class Meta:
         model = Organisation
 
         fields = ["id", "name", "identifier", "year_established"]
 
 
-class OrganisationViewSet(viewsets.ModelViewSet):
-
-    queryset = Organisation.objects.all()
+class PersonSerializer(serializers.ModelSerializer):
 
-    serializer_class = OrganisationSerializer
+    id = serializers.IntegerField(allow_null=True, required=False)
 
-    def perform_create(self, serializer):
-        serializer.save()
+    class Meta:
+        model = Person
+        fields = ["id", "first_name", "last_name", "email", "phone_number"]
 
 
 class ApplicationSerializer(serializers.ModelSerializer):
+
+    contact_person = PersonSerializer(read_only=False, allow_null=True)
+
+    organisation = OrganisationSerializer(read_only=False, allow_null=True)
+
     class Meta:
         model = Application
-        fields = [
-            "id",
-            "description",
-            "reservation_purpose",
-            "organisation",
-            "contact_person",
-            "organisation",
-            "application_period",
-            "user",
-        ]
+        fields = ["id", "organisation", "application_period", "user", "contact_person"]
         read_only_fields = ["user"]
 
-    def validate(self, data):
-        return data
+    def handle_person(self, contact_person_data) -> Union[Person, None]:
+        person = None
+        if contact_person_data is not None:
+            person = Person(**contact_person_data)
+            person.save()
+        return person
+
+    def handle_organisation(self, organisation_data) -> Union[Organisation, None]:
+        organisation = None
+        if organisation_data is not None:
+            organisation = Organisation(**organisation_data)
+            organisation.save()
+        return organisation
+
+    def create(self, validated_data):
```

Comment:
Here is another question, seems like handling updating and creating the nested objects is kind of annoying, is there some easier way to do this. Rather not create new row every time.

(Although i think at least probably contact person and maybe even organisation could be part of the application table itself, but anyway)

Reply:
Maybe we could check in handle functions that if data has ID, then it is and existing object and should be updated rather than created. This kinda seems as clean as it gets to me with nested serializers.

Reply:
Basically that how it works now. If it has an id it'll be updated, if not a new one is created. Maybe could use get_or_create or something to make it clearer?

Reply:
No, actually get_or_create is bit of bad idea, but maybe some other way would be clearer.